SignalMessaging/environment/migrations/OWSDatabaseMigration.m

```diff
@@ -58,6 +58,42 @@ - (void)runUp
         }];
 }
 
+#pragma mark - Database Connections
+
+#ifdef DEBUG
++ (YapDatabaseConnection *)dbReadConnection
+{
+    return self.dbReadWriteConnection;
+}
+
+// Database migrations need to occur _before_ storage is ready (by definition),
+// so we need to use a connection with canWriteBeforeStorageReady set in
+// debug builds.
++ (YapDatabaseConnection *)dbReadWriteConnection
+{
+    static dispatch_once_t onceToken;
+    static YapDatabaseConnection *sharedDBConnection;
+    dispatch_once(&onceToken, ^{
+        sharedDBConnection = [TSStorageManager sharedManager].newDatabaseConnection;
+
+        OWSAssert([sharedDBConnection isKindOfClass:[OWSDatabaseConnection class]]);
+        ((OWSDatabaseConnection *)sharedDBConnection).canWriteBeforeStorageReady = YES;
```

Comment:
There are now asserts that prevent writes before the storage is ready.  We need to make an exception here though.

Reply:
I think the concerns you're raising are already addressed (per the two commits I just pushed), but let's go through this in detail...

* We kick off any version/database migrations after the database registrations.  The migrations' r/w transactions will block on those registrations.  So I don't think we need a separate concept of `StorageExtensionsAreRegistered` since its implicit and guaranteed by the ordering here.
* We now don't mark the app as ready until storage is ready AND the migrations are completed (or skipped).
* We only block read/writes on "storage is ready" (i.e. all db registrations are complete but not all migrations).  We could block read/writes on "app is ready" and make an exception for the db migrations, but I'm not sure it's necessary now that we're being more careful to honor the "app is ready" flag.

> At least one migration is run sync, 

`OWS103EnableVideoCalling` looks like a sync migration, but the core work is a network request which is async.  `OWS104CreateRecipientIdentities` was sync but I've converted it to be async.  

Very happy to continue discussing this.